### PR TITLE
add conditions to check request body

### PR DESCRIPTION
**Pull request by shuoyenl**

### CODE REVIEW

**Review by euforia — CHANGES_REQUESTED** [review]
We need a benchmark on the json body conditions to see the impact on performance with regards to json parsing

**Review by euforia — APPROVED** [review]
We should consider potentially refactoring the match conditions into its own library in the future

File: condition_body_array_match.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		return false
+	}
+	contentType := strings.ToLower(req.Header.Get("Content-type"))
+	switch contentType {
+	case "application/json":
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return false
+		}
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```
Comment (euforia) [inline condition_body_array_match.go:42]: Is the wrapping of the buffer needed here?  You should be able to do without

File: condition_body_match.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		return false
+	}
+	contentType := strings.ToLower(req.Header.Get("Content-type"))
+	switch contentType {
+	case "application/json":
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return false
+		}
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```
Comment (euforia) [inline condition_body_match.go:33]: Is the NopCloser required here?

**Review by euforia — CHANGES_REQUESTED** [review]

**Comment by pavelvasilenka** [comment]
I've done some benchmarking over different condition sets, added in the scope of this PR:
```
$ go test benchmark_condition_test.go -bench=.
Benchmark_String-4                        200000             10469 ns/op
Benchmark_Body-4                           50000             34504 ns/op
Benchmark_BodyArray-4                      50000             33535 ns/op
Benchmark_String_Body-4                    30000             44699 ns/op
Benchmark_String_BodyArray-4               30000             45511 ns/op
Benchmark_Body_BodyArray-4                 20000             65907 ns/op
Benchmark_String_Body_BodyArray-4          20000             79249 ns/op
PASS
ok      command-line-arguments  14.594s
```

File: condition_body_array_match.go (hunk 1 above)
Comment (shuoyenl) [reply condition_body_array_match.go:42]: Request.Body is of type io.ReadCloser and ioutil.NopCloser is to provides a no-op Close method to bytes.Buffer.

**Review by euforia — APPROVED** [review]

**Pull request merged**